api/dbadapters/bigquery.ts

```diff
@@ -199,13 +200,68 @@ export class BigQueryDbAdapter implements IDbAdapter {
       `SELECT * FROM ${CACHED_STATE_TABLE_NAME}`,
       5000 // TODO: Add pagination for 5000+ rows
     );
-    const persistedMetadata = rows.map(row => {
-      const encodedProto = Buffer.from(row.proto, "base64");
+    const persistedMetadata = rows.map((row: IMetadataRow) => {
+      const encodedProto = Buffer.from(row.metadata_proto, "base64");
       return dataform.PersistedTableMetadata.decode(encodedProto);
     });
     return persistedMetadata;
   }
 
+  public async persistStateMetadata(executionGraph: dataform.IExecutionGraph): Promise<void> {
+    const { actions } = executionGraph;
+    const metadataTableCreateQuery = `
+      CREATE TABLE IF NOT EXISTS \`${CACHED_STATE_TABLE_NAME}\` (
+        target_name STRING,
+        metadata_json STRING,
+        metadata_proto STRING
+      )
+    `;
+    await this.runQuery(metadataTableCreateQuery);
+    const tableMetadataMap = new Map<dataform.ITarget, IBigQueryTableMetadata>();
+    await Promise.all(
+      actions.map(async action => {
+        tableMetadataMap.set(action.target, await this.getMetadata(action.target));
+      })
+    );
+    const updateQueries = actions.map(action => {
+      const definitionHash = hashTableDefinition(action);
+      const dependencies = action.dependencyTargets;
+      const metadata =
+        action.type === "operation"
+          ? { lastModifiedTime: `${new Date().valueOf()}` }
+          : tableMetadataMap.get(action.target);
+      const persistTable = dataform.PersistedTableMetadata.create({
+        target: action.target,
+        lastUpdatedMillis: Long.fromString(metadata.lastModifiedTime),
+        definitionHash,
+        dependencies
+      });
+
+      const encodedProtoBuffer = new Buffer(
+        dataform.PersistedTableMetadata.encode(persistTable).finish()
+      );
+
+      const targetName = `${action.target.database}.${action.target.schema}.${action.target.name}`;
+
+      const updateQuery = `MERGE INTO \`${CACHED_STATE_TABLE_NAME}\` T
+          USING (select '${targetName}' as target_name) S
+          ON (T.target_name = S.target_name AND T.target_name = '${targetName}')
+          WHEN NOT MATCHED THEN
+            INSERT (target_name, metadata_json, metadata_proto)
+            VALUES('${targetName}', '${JSON.stringify(
+        persistTable.toJSON()
+      )}', '${encodedProtoBuffer.toString("base64")}')
+          WHEN MATCHED THEN
+            UPDATE SET metadata_json = '${JSON.stringify(persistTable.toJSON())}',
+                metadata_proto = '${encodedProtoBuffer.toString("base64")}'
+          ;`;
+
+      return updateQuery;
+    });
+    const batchQuery = updateQueries.join("\n");
```

Comment:
instead of one `MERGE` statement per table, why not have one `MERGE` statement and have one row in the merged target per table?